clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -244,9 +259,54 @@ public void send(String method, Object... args) throws Exception {
         }
 
         InvocationMessage invocationMessage = new InvocationMessage(method, args);
-        String message = protocol.writeMessage(invocationMessage);
-        logger.log(LogLevel.Debug, "Sending message");
-        transport.send(message);
+        sendHubMessage(invocationMessage);
+    }
+
+    public <T> CompletableFuture<T> invoke(Class<T> returnType, String method, Object... args) throws Exception {
+        InvocationMessage invocationMessage = new InvocationMessage(method, args);
+        String id = connectionState.getNextId();
+        invocationMessage.setInvocationId(id);
+        CompletableFuture<T> future = new CompletableFuture<>();
+
+        connectionState.pendingInvocations.compute(id, (k, v) -> {
+            if (v != null) {
+                // This should never happen
+                throw new IllegalStateException("Invocation Id is already used");
+            }
```

Comment:
If this happens, what state will the connection be in? Will it leak memory? Will things not get cleaned up?

Reply:
> Throws: RuntimeException - or Error if the remappingFunction does so, in which case the mapping is unchanged

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-

So the invocation will never be added and this will throw synchronously without having started any work or added anything to the invocation table.